File: web/src/units/apiClient.ts
import { config } from "./config";

export interface PublicAccount {
  username: string;
  email: string;
  hasOwnKey: boolean;
  isDev: boolean;
  usesDevKey: boolean;
}

export class ApiError extends Error {
  code: string;
  constructor(code: string, message: string) {
    super(message);
    this.name = "ApiError";
    this.code = code;
  }
}

async function call<T>(
  path: string,
  init: RequestInit,
  fetchImpl: typeof fetch = fetch,
): Promise<T> {
  const res = await fetchImpl(`${config.FUNCTIONS_BASE_URL}${path}`, {
    ...init,
    credentials: "include",
    headers: { "content-type": "application/json", ...(init.headers ?? {}) },
  });
  const data = await res.json().catch(() => null);
  if (!res.ok) {
    throw new ApiError(
      data?.error?.code ?? "request_failed",
      data?.error?.message ?? `Request failed (${res.status})`,
    );
  }
  return data as T;
}

export const meApi = {
  async get(fetchImpl?: typeof fetch): Promise<PublicAccount | null> {
    try {
      return await call<PublicAccount>("/me", { method: "GET" }, fetchImpl);
    } catch (e) {
      if (e instanceof ApiError) return null;
      throw e;
    }
  },
};

export const authApi = {
  signup(
    input: { username: string; email: string },
    fetchImpl?: typeof fetch,
  ): Promise<{ pending: boolean }> {
    return call(
      "/auth/signup",
      { method: "POST", body: JSON.stringify(input) },
      fetchImpl,
    );
  },

  request(
    input: { identifier: string },
    fetchImpl?: typeof fetch,
  ): Promise<{ pending: boolean }> {
    return call(
      "/auth/request",
      { method: "POST", body: JSON.stringify(input) },
      fetchImpl,
    );
  },

  verify(
    input: { identifier: string; code: string },
    fetchImpl?: typeof fetch,
  ): Promise<{ account: PublicAccount }> {
    return call(
      "/auth/verify",
      { method: "POST", body: JSON.stringify(input) },
      fetchImpl,
    );
  },

  logout(fetchImpl?: typeof fetch): Promise<{ ok: boolean }> {
    return call("/auth/logout", { method: "POST" }, fetchImpl);
  },
};

export const keyApi = {
  setInitial(
    input: { apiKey: string },
    fetchImpl?: typeof fetch,
  ): Promise<{ ok: boolean }> {
    return call(
      "/key",
      { method: "POST", body: JSON.stringify(input) },
      fetchImpl,
    );
  },

  challenge(fetchImpl?: typeof fetch): Promise<{ pending: boolean }> {
    return call("/key/challenge", { method: "POST" }, fetchImpl);
  },

  reveal(
    input: { code: string },
    fetchImpl?: typeof fetch,
  ): Promise<{ apiKey: string | null; editToken: string }> {
    return call(
      "/key/reveal",
      { method: "POST", body: JSON.stringify(input) },
      fetchImpl,
    );
  },

  edit(
    input: { apiKey: string; editToken: string },
    fetchImpl?: typeof fetch,
  ): Promise<{ ok: boolean }> {
    return call(
      "/key",
      { method: "PUT", body: JSON.stringify(input) },
      fetchImpl,
    );
  },
};

export const allowlistApi = {
  list(fetchImpl?: typeof fetch): Promise<{ emails: string[] }> {
    return call("/dev/allowlist", { method: "GET" }, fetchImpl);
  },

  add(email: string, fetchImpl?: typeof fetch): Promise<{ ok: boolean }> {
    return call(
      "/dev/allowlist",
      { method: "POST", body: JSON.stringify({ email }) },
      fetchImpl,
    );
  },

  remove(email: string, fetchImpl?: typeof fetch): Promise<{ ok: boolean }> {
    return call(
      "/dev/allowlist",
      { method: "DELETE", body: JSON.stringify({ email }) },
      fetchImpl,
    );
  },
};

File: ios/FaceBackKit/Sources/FaceBackKit/APIClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// The backend surface the native app calls. Mirrors the MVP subset of
/// `web/src/units/apiClient.ts` and `generationClient.ts`.
public protocol FaceBackAPI {
    func me() async -> PublicAccount?
    func signup(username: String, email: String) async throws
    func requestCode(identifier: String) async throws
    func verify(identifier: String, code: String) async throws -> PublicAccount
    func logout() async throws
    func setInitialKey(_ apiKey: String) async throws
    func generate(_ image: ImagePayload) async throws -> ImagePayload
}

/// Talks to the Cloudflare backend over an injectable `HTTPTransport`. The
/// `fb_session` cookie is carried automatically by the underlying `URLSession`
/// (see `URLSessionTransport`), reproducing the web client's `credentials: "include"`.
public final class APIClient: FaceBackAPI {
    private let baseURL: URL
    private let transport: HTTPTransport

    public init(baseURL: URL, transport: HTTPTransport) {
        self.baseURL = baseURL
        self.transport = transport
    }

    public convenience init(baseURL: URL, session: URLSession = .shared) {
        self.init(baseURL: baseURL, transport: URLSessionTransport(session: session))
    }

    // MARK: - Endpoints

    public func me() async -> PublicAccount? {
        do {
            let data = try await perform("/me", method: "GET", body: nil)
            return try JSONDecoder().decode(PublicAccount.self, from: data)
        } catch {
            return nil
        }
    }

    public func signup(username: String, email: String) async throws {
        _ = try await perform("/auth/signup", method: "POST",
                              body: jsonBody(["username": username, "email": email]))
    }

    public func requestCode(identifier: String) async throws {
        _ = try await perform("/auth/request", method: "POST",
                              body: jsonBody(["identifier": identifier]))
    }

    public func verify(identifier: String, code: String) async throws -> PublicAccount {
        let data = try await perform("/auth/verify", method: "POST",
                                     body: jsonBody(["identifier": identifier, "code": code]))
        return try JSONDecoder().decode(AccountEnvelope.self, from: data).account
    }

    public func logout() async throws {
        _ = try await perform("/auth/logout", method: "POST", body: nil)
    }

    public func setInitialKey(_ apiKey: String) async throws {
        _ = try await perform("/key", method: "POST", body: jsonBody(["apiKey": apiKey]))
    }

    public func generate(_ image: ImagePayload) async throws -> ImagePayload {
        let data = try await perform("/generate", method: "POST",
                                     body: try? JSONEncoder().encode(GenerateRequest(image: image)))
        return try JSONDecoder().decode(GenerateResponse.self, from: data).image
    }

    // MARK: - Core

    private func perform(_ path: String, method: String, body: Data?) async throws -> Data {
        guard let url = URL(string: baseURL.absoluteString + path) else {
            throw APIError(code: "bad_url", message: "Invalid URL for \(path)")
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = body

        let (data, status) = try await transport.send(request)
        guard (200..<300).contains(status) else {
            if let envelope = try? JSONDecoder().decode(ServerErrorEnvelope.self, from: data) {
                throw APIError(code: envelope.error.code, message: envelope.error.message)
            }
            throw APIError(code: "request_failed", message: "Request failed (\(status))")
        }
        return data
    }

    private func jsonBody(_ dict: [String: String]) -> Data? {
        try? JSONEncoder().encode(dict)
    }
}

private struct AccountEnvelope: Decodable { let account: PublicAccount }
private struct GenerateRequest: Encodable { let image: ImagePayload }
private struct GenerateResponse: Decodable { let image: ImagePayload }
